pinot-core/src/main/java/com/linkedin/pinot/core/query/scheduler/MultiLevelPriorityQueue.java

```diff
@@ -0,0 +1,231 @@
+    queryDeadlineMillis = config.getInt(QUERY_DEADLINE_SECONDS_KEY, 30) * 1000;
+    wakeUpTimeMicros = config.getInt(QUEUE_WAKEUP_MICROS, DEFAULT_WAKEUP_MICROS);
+    maxPendingPerGroup = config.getInt(MAX_PENDING_PER_GROUP_KEY, 10);
+    this.config = config;
+    this.resourceManager = resourceManager;
+    this.groupFactory = groupFactory;
+    this.groupSelector = groupMapper;
+  }
+
+  @Override
+  public void put(@Nonnull SchedulerQueryContext query) throws OutOfCapacityError {
+    Preconditions.checkNotNull(query);
+    queueLock.lock();
+    String groupName = groupSelector.getSchedulerGroupName(query);
+    try {
+      SchedulerGroup groupContext = getOrCreateGroupContext(groupName);
+      checkGroupHasCapacity(groupContext);
+      query.setSchedulerGroupContext(groupContext);
+      groupContext.addLast(query);
+      queryReaderCondition.signal();
+    } finally {
+      queueLock.unlock();
+    }
+  }
+
+  /**
+   * Blocking call to read the next query in order of priority
+   * @return
+   */
+  @Override
+  public @Nullable SchedulerQueryContext take() {
+    queueLock.lock();
+    try {
+      while (true) {
+        SchedulerQueryContext schedulerQueryContext;
+        while ( (schedulerQueryContext = takeNextInternal()) == null) {
+          try {
+            queryReaderCondition.await(wakeUpTimeMicros, TimeUnit.MICROSECONDS);
+          } catch (InterruptedException e) {
+            return null;
+          }
+        }
+        return schedulerQueryContext;
+      }
+    } finally {
+      queueLock.unlock();
+    }
+  }
+
+  @Nonnull
+  @Override
+  public List<SchedulerQueryContext> drain() {
+    List<SchedulerQueryContext> pending = new ArrayList<>();
+    queueLock.lock();
+    try {
+      for (Map.Entry<String, SchedulerGroup> groupEntry : schedulerGroups.entrySet()) {
+        SchedulerGroup group = groupEntry.getValue();
+        while (!group.isEmpty()) {
+          pending.add(group.removeFirst());
+        }
+      }
+    } finally {
+      queueLock.unlock();
+    }
+    return pending;
+  }
+
+  private SchedulerQueryContext takeNextInternal() {
+    SchedulerGroup currentWinnerGroup = null;
+    long startTime = System.nanoTime();
+    StringBuilder sb = new StringBuilder("SchedulerInfo:");
+    long deadlineEpochMillis = currentTimeMillis() - queryDeadlineMillis;
+    for (Map.Entry<String, SchedulerGroup> groupInfoEntry : schedulerGroups.entrySet()) {
+      SchedulerGroup group = groupInfoEntry.getValue();
+      sb.append(group.toString());
+      group.trimExpired(deadlineEpochMillis);
+      if (group.isEmpty() || !resourceManager.canSchedule(group)) {
+        continue;
+      }
+
+      if (currentWinnerGroup == null) {
+        currentWinnerGroup = group;
+        continue;
+      }
+
+      // Preconditions:
+      // a. currentGroupResources <= hardLimit
+      // b. selectedGroupResources <= hardLimit
+      // We prefer group with higher tokens but with resource limits.
```

Comment:
Since this class may deal with token-based or fcfs scheduler, shouldn't these comments go to another file? (or, be more generic here?)